examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,637 @@
+# sub-estimator, or a consumer starts consuming a metadata which it wasn't in
+# an older release. In this case, a warning should be raised for a while, to
+# let users know the behavior is changed from previous versions.
+
+
+class MetaRegressor(MetaEstimatorMixin, RegressorMixin, BaseEstimator):
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def fit(self, X, y, **fit_params):
+        params = process_routing(self, "fit", fit_params)
+        self.estimator_ = clone(self.estimator).fit(X, y, **params.estimator.fit)
+
+    def get_metadata_routing(self):
+        router = MetadataRouter(owner=self.__class__.__name__).add(
+            estimator=self.estimator, method_mapping="one-to-one"
+        )
+        return router
+
+
+# %%
+# As explained above, this is now a valid usage:
+
+reg = MetaRegressor(estimator=LinearRegression().set_fit_request(sample_weight=True))
+reg.fit(X, y, sample_weight=my_weights)
+
+
+# %%
+# Now imagine we further develop ``MetaRegressor`` and it now also *consumes*
+# ``sample_weight``:
+
+
+class WeightedMetaRegressor(MetaEstimatorMixin, RegressorMixin, BaseEstimator):
+    __metadata_request__fit = {"sample_weight": RequestType.WARN}
+
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def fit(self, X, y, sample_weight=None, **fit_params):
+        params = process_routing(self, "fit", fit_params, sample_weight=sample_weight)
+        check_metadata(self, sample_weight=sample_weight)
+        self.estimator_ = clone(self.estimator).fit(X, y, **params.estimator.fit)
+
+    def get_metadata_routing(self):
+        router = (
+            MetadataRouter(owner=self.__class__.__name__)
+            .add_self(self)
+            .add(estimator=self.estimator, method_mapping="one-to-one")
+        )
+        return router
+
+
+# %%
+# The above implementation is almost no different than ``MetaRegressor``, and
+# because of the default request value defined in ``__metadata_request__fit``
+# there is a warning raised.
+
+with warnings.catch_warnings(record=True) as record:
+    WeightedMetaRegressor(
+        estimator=LinearRegression().set_fit_request(sample_weight=False)
+    ).fit(X, y, sample_weight=my_weights)
+for w in record:
+    print(w.message)
+
+
+# %%
+# When an estimator supports a metadata which wasn't supported before, the
+# following pattern can be used to warn the users about it.
+
+
+class ExampleRegressor(RegressorMixin, BaseEstimator):
+    __metadata_request__fit = {"sample_weight": RequestType.WARN}
+
+    def fit(self, X, y, sample_weight=None):
+        check_metadata(self, sample_weight=sample_weight)
+        return self
+
+    def predict(self, X):
+        return np.zeros(shape=(len(X)))
+
+
+with warnings.catch_warnings(record=True) as record:
+    MetaRegressor(estimator=ExampleRegressor()).fit(X, y, sample_weight=my_weights)
+for w in record:
+    print(w.message)
+
+# %%
+# Deprecation to Give Users Time to Adapt their Code
```

Comment:
The code below just raise an error but it does not allow users to adapt their code, isn't it?

Reply:
ah yeah, we removed this part, this should have been removed as well.